src/cursor/cur_version.c

```diff
@@ -0,0 +1,611 @@
+/*
+ * __curversion_set_key --
+ *     WT_CURSOR->set_key implementation for version cursors.
+ */
+static void
+__curversion_set_key(WT_CURSOR *cursor, ...)
+{
+    WT_CURSOR *table_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+    uint32_t flags;
+    va_list ap;
+
+    session = CUR2S(cursor);
+
+    /* Reset the cursor every time for a new key. */
+    if ((ret = cursor->reset(cursor)) != 0)
+        WT_IGNORE_RET(__wt_panic(session, ret, "failed to reset cursor"));
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    va_start(ap, cursor);
+    flags = table_cursor->flags;
+    /* Pass on the raw flag. */
+    if (F_ISSET(cursor, WT_CURSTD_RAW))
+        flags |= WT_CURSTD_RAW;
+    if ((ret = __wt_cursor_set_keyv(table_cursor, flags, ap)) != 0)
+        WT_IGNORE_RET(__wt_panic(session, ret, "failed to set key"));
+    va_end(ap);
+}
+
+/*
+ * __curversion_get_key --
+ *     WT_CURSOR->get_key implementation for version cursors.
+ */
+static int
+__curversion_get_key(WT_CURSOR *cursor, ...)
+{
+    WT_CURSOR *table_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    uint32_t flags;
+    va_list ap;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    va_start(ap, cursor);
+    flags = table_cursor->flags;
+    /* Pass on the raw flag. */
+    if (F_ISSET(cursor, WT_CURSTD_RAW))
+        flags |= WT_CURSTD_RAW;
+    WT_ERR(__wt_cursor_get_keyv(table_cursor, flags, ap));
+
+err:
+    va_end(ap);
+    return (ret);
+}
+
+/*
+ * __curversion_get_value --
+ *     WT_CURSOR->get_value implementation for version cursors.
+ */
+static int
+__curversion_get_value(WT_CURSOR *cursor, ...)
+{
+    WT_CURSOR *table_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_ITEM(data);
+    WT_DECL_ITEM(metadata);
+    WT_DECL_PACK_VALUE(pv);
+    WT_DECL_RET;
+    WT_PACK pack;
+    WT_SESSION_IMPL *session;
+    const uint8_t *p, *end;
+    va_list ap;
+
+    CURSOR_API_CALL(cursor, session, get_value, NULL);
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    va_start(ap, cursor);
+    if (F_ISSET(cursor, WT_CURSTD_RAW)) {
+        /* Extract metadata and value separately as raw data. */
+        metadata = va_arg(ap, WT_ITEM *);
+        metadata->data = cursor->value.data;
+        metadata->size = cursor->value.size;
+        data = va_arg(ap, WT_ITEM *);
+        data->data = table_cursor->value.data;
+        data->size = table_cursor->value.size;
+    } else {
+        /*
+         * Unpack the metadata. We cannot use the standard get value function here because variable
+         * arguments cannot be partially extracted by different function calls.
+         */
+        p = (uint8_t *)cursor->value.data;
+        end = p + cursor->value.size;
```

Comment:
Can `cursor.value.data` be NULL in this path? If so, you'll hit the V4 sanitizer checks when you try and add `size` to it.